### ISSUE TITLE
Validator: improve slice validation (nested slicing)

### ISSUE THREAD

**Issue by ArdonToonstra** [FirelyTeam/fhir-net-api#648]
As discussed today, we encounter problems validating instances of our zib-bloodpressure profile. This is a derived profile from vitalsigns bloodpressure which is a derived profile from the vitalsigns profile. Please find all profiles and instances attached.

Validator goes wrong on the sliced .components and sliced code.codings. In our original profile, the discriminator is 'open'. In the attached profile, this is set to 'closed'. See the image for 

As discussed as well, the attached profiles and instances provide a good test set for improving the slice validation.

[validator test bloodpressure profile.zip](https://github.com/ewoutkramer/fhir-net-api/files/2227300/validator.test.bloodpressure.profile.zip)

**Comment by ArdonToonstra** [comment]
Validator can't handle nested slices. We removed the nested slices for test purposes, which resulted in a successful validation of our instance.

**Comment by ewoutkramer** [comment]
Will focus on this in the next release - the slice/discriminator system needs rework.

**Comment by dnperfors** [comment]
Any progress on this? I still get warnings that slices are not supported...

**Comment by ArdonToonstra** [comment]
@marcovisserFurore ,

I don't believe nested slicing works. Please find attached my simplified testset that contains one profile on Observation and 1 example. The Observation.component is sliced, and the component.code.coding is sliced. (nested slicing)

I get the following error code:

`[FATAL] Internal logic failure: Child with name 'coding' is sliced to more than one choice and cannot be used as a discriminator at 'http://example.org/fhir/StructureDefinition/MyObservation#Observation.component.code'  (at Observation)`

**Comment by ArdonToonstra** [comment]
[Nested slicing.zip](https://github.com/FirelyTeam/fhir-net-api/files/3717696/Nested.slicing.zip)

**Comment by marcovisserFurore** [comment]
Thanks for testing this. I will reopen the issue and investigate this later.

**Comment by marcovisserFurore** [comment]
@wmrutten: can we have a look at this example? 

If do understand correctly, there is a value slice on element `Observation.component` on value `Observation.component.code.coding,code`. Futhermore there is a (nested) value slice on `Observation.component.code.coding` on value `Observation.component.code.coding.system`.

The validator responds with this message:

```
[FATAL] Internal logic failure: Child with name 'coding' is sliced to more than one choice and cannot be used as a discriminator at 'http://example.org/fhir/StructureDefinition/MyObservation#Observation.component.code'  (at Observation)
``` 

I want to make sure if that is alright.

**Comment by ewoutkramer** [comment]
This is the expected behaviour, although it may not be what @ArdonToonstra was hoping.  The problem is that the validator looks at the first level of slicing (the components) where it sees that the components are discriminated based on `code.coding.code`.  It then goes out into the deeper constraints to find the fixed code for both component slices. Unfortunately, when looking for a fixed code to discriminate the first slice of component (`sliceComponent-01`), it finds two such codes: `SNOMED-CODE` and `LOINC-CODE`. This is due to the fact that within `sliceComponent-01`, there is a nested slicing (sliceCoding-SNOMED, sliceCoding-LOINC) _which has different values for the discriminator of the parent slice_. 

The validator detects this fact and reports that it cannot use `code.coding.code` as a discriminator for the parent slice, since the children do not actually fix this code to a single value, but to multiple values.

I guess @ArdonToonstra is expecting here that *any* of the fixed values introduced by the nested slicing will be accepted as _the_ fixed value for the parent's discriminator? 

For now, this means that you can certainly nested slices, but these slices cannot vary an element used by a parent slice as the discriminator. We might need to discuss whether this is desired and expected or not.

**Comment by joelemaltais** [comment]
Hi there, so it seems that the validator does not validate any FHIR element sliced by profile. Will this be supported in the next API release? Ran into this issue in a large project and have not figured a way around it. We switched of validation for now, meaning we can't go on to  production and don't get client acceptance test = get paid till we figure something out. 
@alexzautke knows the details.
